File: database-service/src/main/java/com/bnorm/barkeep/db/DbBookService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db;

import java.util.List;

import javax.persistence.EntityManager;

import com.bnorm.barkeep.model.Book;
import com.bnorm.barkeep.model.Recipe;
import com.bnorm.barkeep.service.BookService;

public class DbBookService extends AbstractDbService implements BookService {

  public DbBookService(EntityManager entityManager) {
    super(entityManager);
  }

  @Override
  public List<Book> getBooks() {
    return super.getBooks();
  }

  @Override
  public BookEntity getBook(long bookId) {
    return super.getBook(bookId);
  }

  @Override
  public BookEntity createBook(Book book) {
    if (book.getId() != null) {
      throw new IllegalArgumentException(String.format("Cannot create book that already has an id=%d", book.getId()));
    } else if (book.getOwner() == null) {
      throw new IllegalArgumentException("Cannot create book without an owner");
    }
    UserEntity userEntity = find(book.getOwner());
    if (userEntity == null) {
      throw new IllegalArgumentException(String.format("Cannot create book with an unknown owner id=%d",
                                                       book.getOwner().getId()));
    }

    return transaction(em -> {
      BookEntity bookEntity = new BookEntity();

      bookEntity.setTitle(book.getTitle());
      bookEntity.setDescription(book.getDescription());
      bookEntity.setOwner(userEntity);
      if (book.getRecipes() != null) {
        for (Recipe recipe : book.getRecipes()) {
          bookEntity.addRecipe(find(recipe));
        }
      }

      em.persist(bookEntity);

      userEntity.addBook(bookEntity);
      return bookEntity;
    });
  }

  @Override
  public BookEntity setBook(long bookId, Book book) {
    BookEntity bookEntity = findBook(bookId);
    requireExists(bookEntity, bookId, "book");
    requireMatch(book, bookId, "book");

    return transaction(em -> {
      if (book.getTitle() != null) {
        bookEntity.setTitle(book.getTitle());
      }
      if (book.getDescription() != null) {
        bookEntity.setDescription(book.getTitle());
      }
      if (book.getOwner() != null) {
        bookEntity.setOwner(find(book.getOwner()));
      }
      if (book.getRecipes() != null) {
        // todo is this correct?
        for (Recipe recipe : book.getRecipes()) {
          bookEntity.addRecipe(find(recipe));
        }
      }

      return bookEntity;
    });
  }

  @Override
  public void deleteBook(long bookId) {
    transaction(em -> {
      BookEntity bookEntity = findBook(bookId);
      requireExists(bookEntity, bookId, "book");
      em.remove(bookEntity);
    });
  }
}

File: service-database/src/main/kotlin/com/bnorm/barkeep/db/DbBookService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db

import com.bnorm.barkeep.model.Book
import com.bnorm.barkeep.model.BookSpec
import com.bnorm.barkeep.service.BookService
import javax.persistence.EntityManager

class DbBookService(private val em: EntityManager,
                    private val recipeService: DbRecipeService,
                    private val userService: DbUserService) : BookService {

  override fun getBooks(): List<Book> {
    return em.createNamedQuery("BookEntity.findAll", BookEntity::class.java).resultList
  }

  fun findBook(id: Long): BookEntity? {
    return em.find(BookEntity::class.java, id)
  }

  override fun getBook(id: Long): BookEntity {
    return findBook(id) ?: throw IllegalArgumentException("Cannot find book with id=$id")
  }

  override fun createBook(book: BookSpec): BookEntity {
    val owner = book.owner ?: throw IllegalArgumentException("Cannot create book without an owner")
    val userEntity = userService.findUser(owner.id) ?: throw IllegalArgumentException("Cannot create book with an unknown owner id=${owner.id}")

    val bookEntity = BookEntity()
    bookEntity.title = book.title
    bookEntity.description = book.description
    bookEntity.owner = userEntity
    bookEntity.addUser(userEntity)

    em.txn {
      em.persist(bookEntity)
      userEntity.addBook(bookEntity)
    }
    return bookEntity
  }

  override fun addBookRecipe(bookId: Long, recipeId: Long) {
    val bookEntity = getBook(bookId)
    val recipeEntity = recipeService.getRecipe(recipeId)

    em.txn {
      bookEntity.addRecipe(recipeEntity)
    }
  }

  override fun removeBookRecipe(bookId: Long, recipeId: Long) {
    val bookEntity = getBook(bookId)
    val recipeEntity = recipeService.getRecipe(recipeId)

    em.txn {
      bookEntity.removeRecipe(recipeEntity)
    }
  }

  override fun setBook(id: Long, book: BookSpec): BookEntity {
    val bookEntity = getBook(id)

    em.txn {
      book.title?.apply { bookEntity.title = this }
      book.description?.apply { bookEntity.description = this }
      book.owner?.apply { bookEntity.owner = userService.getUser(this.id) }
    }
    return bookEntity
  }

  override fun deleteBook(id: Long) {
    em.txn {
      em.remove(getBook(id))
    }
  }
}
